python_coreml_stable_diffusion/activation_quantization.py

```diff
@@ -0,0 +1,489 @@
+        timestep = args[1]
+        if len(timestep.shape) == 0:
+            timestep = timestep[None]
+        timestep = timestep.expand(sample.shape[0])
+        encoder_hidden_states = kwargs["encoder_hidden_states"]
+        encoder_hidden_states = encoder_hidden_states.permute((0, 2, 1)).unsqueeze(2)
+        modified_args = (sample, timestep, encoder_hidden_states)
+        return (modified_args, {})
+
+    return pipe.unet.register_forward_pre_hook(hook, with_kwargs=True)
+
+def prepare_pipe(pipe, unet):
+    """
+    Create a new pipeline from `pipe` with `unet` as the noise predictor
+    """
+    new_pipe = deepcopy(pipe)
+    unet.to(new_pipe.unet.device)
+    new_pipe.unet = unet
+    pre_hook_handle = register_input_preprocessing_hook(new_pipe)
+    return new_pipe, pre_hook_handle
+
+def run_pipe(pipe):
+    gen = torch.manual_seed(args.seed)
+    kwargs = dict(
+        prompt=RANDOM_TEST_DATA,
+        output_type="latent",
+        generator=gen,
+    )
+    return np.array([latent.cpu().numpy() for latent in pipe(**kwargs).images])
+
+
+def get_reference_pipeline(model_version):
+    # Initialize pipe
+    pipe = StableDiffusionPipeline.from_pretrained(
+        model_version,
+        use_safetensors=True,
+        use_auth_token=True,
+    )
+    DEFAULT_NUM_INFERENCE_STEPS = 50
+    pipe.scheduler.set_timesteps(DEFAULT_NUM_INFERENCE_STEPS)
+
+    # Initialize reference unet
+    unet_cls = unet.UNet2DConditionModel
+    reference_unet = unet_cls(**pipe.unet.config).eval()
+    reference_unet.load_state_dict(pipe.unet.state_dict())
+
+    # Initialize reference pipeline
+    ref_pipe, _ = prepare_pipe(pipe, reference_unet)
+
+    del pipe
+    gc.collect()
+    return ref_pipe
+
+def main(args):
+    # Initialize reference pipeline
+    ref_pipe = get_reference_pipeline(args.model_version)
+    if torch.cuda.is_available():
+        device = "cuda"
+    else:
+        device = "cpu"
+    logger.debug(f"Placing pipe in {device}")
+    ref_pipe.to(device)
+    # Generate baseline outputs
+    ref_out = run_pipe(ref_pipe)
+
+    # Setup artifact file paths
+    os.makedirs(args.o, exist_ok=True)
+    recipe_json_path = os.path.join(args.o, f"{args.model_version.replace('/', '_')}_quantization_recipe.json")
+    calibration_dir = os.path.join(args.o, f"calibration_data_{args.model_version.replace('/', '_')}")
+
+    # Generate calibration data 
+    if args.generate_calibration_data:
+        generate_calibration_data(ref_pipe, args, calibration_dir)
+
+    # Compute layer-wise PSNR
+    if args.layerwise_sensitivity:
+        logger.info("Compute Layer-wise PSNR")
+        quantizable_modules = get_quantizable_modules(ref_pipe.unet)
+
+        results = {
+            'conv': {},
+            'einsum': {},
+            'model_version': args.model_version
+        }
+        dataloader = unet_data_loader(calibration_dir, device, args.calibration_nsamples)
+
+        for module_type, module_name in quantizable_modules:
```

Comment:
Since this loop takes hrs to run (?), maybe would make sense to use `tqdm` here to log/print % progress